### PR TITLE
F4BY HAL

### PR DESCRIPTION
**Pull request by mstrinzha**
Support for F4BY board. More detail are here: http://diydrones.com/forum/topics/f4by-controller-with-arducopter-3-3-2-arduplane-3-4-tested

### CODE REVIEW

File: libraries/AP_HAL_F4BY/AP_HAL_F4BY_Namespace.h [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+	class F4BYAnalogIn;
+	class F4BYAnalogSource;
+	class F4BYUtil;
+        class F4BYGPIO;
+        class F4BYDigitalSource;
+        class NSHShellStream;
+        class F4BYI2CDriver;
+        class F4BY_I2C;
+        class Semaphore;
+}
```
Comment (guludo) [inline libraries/AP_HAL_F4BY/AP_HAL_F4BY_Namespace.h]: Declaring classes this way isn't really the best approach. See https://github.com/ArduPilot/ardupilot/pull/4574

**Comment by guludo** [comment]
This is missing the build integration in Waf. And I think we should skip the integration in the make-based build system, since it's been deprecated.

**Comment by guludo** [comment]
Oh, and regarding the integration with Waf, I think following https://github.com/ArduPilot/ardupilot/pull/4647 may help you.

File: libraries/AP_HAL_F4BY/AnalogIn.cpp [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+    uint8_t pin;
+    float scaling;
+} pin_scaling[] = {
+    // F4BY has 4 FMU analog input pins
+    { 10, (5.7f*3.3f)/4096 }, // FMU battery on multi-connector pin 5,
+                            // 5.7:1 scaling
+    { 11,  6.6f/4096  }, // analog airspeed input, 2:1 scaling
+    { 12,  3.3f/4096  }, // analog2, on SPI port pin 3
+    { 13, 16.8f/4096  }, // analog3, on SPI port pin 4
+};
```
Comment (magicrub) [inline libraries/AP_HAL_F4BY/AnalogIn.cpp]: fixed analog input scaling tied to certain sensors? That's no good.

**Comment by magicrub** [comment]
What happened to the days where someone makes a pixhawk clone and it takes either zero or only a couple lines of code to support it? With this and the PX4Firmware PR it's a total of 11k LOC in 3 commits.

**Comment by OXINARF** [comment]
> And I think we should skip the integration in the make-based build system, since it's been deprecated

It's only deprecated for Linux

> What happened to the days where someone makes a pixhawk clone and it takes either zero or only a couple lines of code to support it?

This isn't a Pixhawk clone, that's why it doesn't work without modifications. But this can be substantially reduced since the HAL is actually very similar to the PX4 one (and VRBRAIN one).

@tridge Should the PX4 HAL renaming be discussed in a dev call?

**Comment by guludo** [comment]
> It's only deprecated for Linux

Oops

**Comment by magicrub** [comment]
Sorry, didn't really mean to call it a Pixahwk clone. I meant why didn't it just take advantage of the HAL so the change footprint was tiny. It handles sensor IC rotations, it handles SPI/I2C bus transactions pretty generically. @mstrinzha is there something in the PX4Firmware change that *needs* to be there? Have you looked into simply adding a HAL module in ArduPilot? The VRBrain should try and get itself merged into the HAL_PX4 folder at some point. They have a few Nuttx changes that are nice but I'm not sure if it warrants a whole new platform layer. That's another topic though

**Comment by magicrub** [comment]
@mstrinzha Typically wen we add a new board you become a maintainer of it in the code. Are you able to reply to any of the above comments? We'd like this in but it needs some changes. Your feedback would be appreciated

File: libraries/AP_HAL_F4BY/AP_HAL_F4BY_Namespace.h (hunk 1 above)
Comment (dipspb) [inline libraries/AP_HAL_F4BY/AP_HAL_F4BY_Namespace.h]: Could you be so kind to provide more details on what is wrong here? It's just a forward declaration of classes

Comment (guludo) [inline libraries/AP_HAL_F4BY/AP_HAL_F4BY_Namespace.h]: This works, but every time this header is changed, it triggers unnecessary rebuild for a lot of cpp files. Also, see this commit, which has already been merged, https://github.com/ArduPilot/ardupilot/pull/4574/commits/f36e8d9c0559ae7214a9d4c5c9b1ca06209ba390.

File: libraries/AP_HAL_F4BY/AnalogIn.cpp (hunk 2 above)
Comment (magicrub) [inline libraries/AP_HAL_F4BY/AnalogIn.cpp]: I understand now. Yes, these are fine.

**Comment by magicrub** [comment]
I really like that this doesn't touch module/PX4Firmware :)

**Comment by OXINARF** [comment]
@magicrub It does: https://github.com/ArduPilot/PX4Firmware/pull/65

This PR will need to reworked after we have changed HAL_PX4 to HAL_Nuttx and merged HAL_VRBrain.

File: libraries/AP_HAL_F4BY/AP_HAL_F4BY_Namespace.h (hunk 1 above)
Comment (dipspb) [inline libraries/AP_HAL_F4BY/AP_HAL_F4BY_Namespace.h]: We preparing new PR for F4BY HAL and intended to delete this file and to rename classes, but I checked it once more and just discovered there are a lot of places in master branch with a such kind of approach implemented. For example:

https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_HAL/AP_HAL_Namespace.h
https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_HAL_Empty/AP_HAL_Empty_Namespace.h
https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_HAL_PX4/AP_HAL_PX4_Namespace.h
https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_HAL_QURT/AP_HAL_QURT_Namespace.h
https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_HAL_SITL/AP_HAL_SITL_Namespace.h
https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_HAL_VRBRAIN/AP_HAL_VRBRAIN_Namespace.h

Are you sure we still need to refactor it?

Comment (OXINARF) [inline libraries/AP_HAL_F4BY/AP_HAL_F4BY_Namespace.h]: Yes, other HALs haven't been converted, but like Gustavo said, it is good practice to do it like Linux has now since it will reduce any necessary recompilation.

By the way, I'm not sure what you are doing but I hope you aren't making a full new HAL very similar to PX4 like this one did.

Comment (dipspb) [inline libraries/AP_HAL_F4BY/AP_HAL_F4BY_Namespace.h]: Thank you, Francisco. We have a fix for namespaces already so I just checked once more we need to do that.

That's for "_full new HAL very similar to PX4_" I'm not quite sure I can see your point. Would you be so kind to clarify it? We implemented it as a separate HAL due to hardware difference in F4BY. For example, we have no i/o processor and there other differences also. In case a full new HAL doesn't seems as reasonable approach for this case what do you see as a more reasonable way?

Comment (OXINARF) [inline libraries/AP_HAL_F4BY/AP_HAL_F4BY_Namespace.h]: If you look at this PR you'll see the main reason it wasn't merged is that it basically copies the PX4 HAL instead of adapting it. If you are using PX4Firmware there is no reason to have a separate HAL in ArduPilot, it is just a different board on the same abstraction layer.

**Comment by mstrinzha** [comment]
PR is reworked and ready to be merged.

File: libraries/AP_RPM/RPM_Pin.cpp [hunk 3]
```diff
@@ -90,9 +90,11 @@ void AP_RPM_Pin::update(void)
         case 54:
             gpio = GPIO_GPIO4_INPUT;
             break;
+#if defined(GPIO_GPIO5_INPUT)
```
Comment (khancyr) [inline libraries/AP_RPM/RPM_Pin.cpp]: why this change?

Comment (mstrinzha) [inline libraries/AP_RPM/RPM_Pin.cpp]: We have only 5 inputs.

**Comment by mstrinzha** [comment]
All checks fail because PX4Firmware does not contain https://github.com/ArduPilot/PX4Firmware/pull/65

**Comment by mstrinzha** [comment]
@OXINARF, @magicrub, @guludo, @khancyr, @rmackay9 please take a look at this PR.

**Comment by davidbuzz** [comment]
this PR is 2 years old and has not been looked at for over  6 months, and in the interveening time the ChibiOS port is becoming a replacement for the old PX4/NuttX layer.   I don't think this PR has any chance of being merged as-is.    I welcome a new pull request for the F4BY HAL based on recent HAL work, but I'm closing this one in the mean time.
